de.dlr.sc.virsat.model.edit/src/de/dlr/sc/virsat/model/concept/types/structural/level/HierarchyLevelChecker.java

```diff
@@ -0,0 +1,368 @@
+
+import de.dlr.sc.virsat.model.concept.types.structural.IBeanStructuralElementInstance;
+import de.dlr.sc.virsat.model.concept.types.structural.tree.BeanStructuralTreeTraverser;
+
+/**
+ * <pre>
+ * A class to check elements against a custom tree level structure which are
+ * specified by implementations of {@link IHierarchyLevel}s
+ * 
+ * An example of such a tree structure is a model that consists of a SYSTEM,
+ * SUBSYTEM and EQUIPMENT levels These levels are implemented with the
+ * {@link IHierarchyLevel} interface
+ * 
+ * System MySatellite
+ *  |
+ *  ----Subsystem AOCS
+ *  |	|
+ *  |	----Equipment ReactionWheel1
+ *  |	----Equipment ReactionWheel2
+ *  |
+ *  ----NewElement    <====  This class can check what level this element might be on
+ *  	|
+ *  	----Equipment Thruster
+ * 
+ * </pre>
+ * 
+ * 
+ */
+public class HierarchyLevelChecker {
+
+	protected List<IHierarchyLevel> levels;
+
+	/**
+	 * @param levels
+	 *            A list that specifies the concrete order of tree levels.
+	 * 
+	 *            Levels should not overlap - for any element only one level can
+	 *            have true as result from the
+	 *            {@link IHierarchyLevel#isOnLevel(IBeanStructuralElementInstance)} method.
+	 *            Otherwise an {@link IllegalArgumentException} is thrown
+	 */
+	public HierarchyLevelChecker(List<IHierarchyLevel> levels) {
+		this.levels = levels;
+	}
+
+	/**
+	 * Checks which levels an element can be on
+	 * 
+	 * @param bean
+	 *            the structural element bean to check
+	 * @return a set of applicable levels from the list of levels passed in the
+	 *         constructor
+	 */
+	public Set<IHierarchyLevel> getApplicableLevels(IBeanStructuralElementInstance bean) {
+
+		// Check if element belongs to level already - then it can by definition of this
+		// checker only be on this level
+		IHierarchyLevel currentLevel = getLevelOfBean(bean);
+		if (currentLevel != null) {
+			return Collections.singleton(currentLevel);
+		}
+
+		return deduceApplicableLevels(bean);
+	}
+
+	/**
+	 * Deduces applicable levels based on the levels already present in the tree
+	 * @param bean
+	 *            the structural element bean to check
+	 * @return set of applicable levels
+	 */
+	protected Set<IHierarchyLevel> deduceApplicableLevels(IBeanStructuralElementInstance bean) {
+		Set<IHierarchyLevel> applicableLevels = new HashSet<>();
+		int minLevelIndex = 0;
+		int maxLevelIndex = levels.size() - 1;
+		
+		// Get first parent with level
+		IBeanStructuralElementInstance parent = getFirstParentWithLevel(bean);
+		// Get children with level
+		List<IBeanStructuralElementInstance> children = getChildrenWithLevel(bean);
+
+		// Check maximum and minimum level index
+		if (parent != null) {
+			int minLevelFromParent = getMinIndexFromParentLevel(parent);
+			int maxLevelFromParent = getMaxIndexFromTreeDistanceOfParent(parent, bean);
+
+			if (minLevelFromParent > minLevelIndex) {
+				minLevelIndex = minLevelFromParent;

```

Comment:
`minLevelIndex = Math.max(minLevelIndex, minLevelFromParent)` for simplification